### ISSUE TITLE
About short-history ablation

### ISSUE THREAD

**Issue by y-ito-1985** [yulincho/MTIL#1]
Thank you for your excellent work.
I think that taking long-term context into account is a significant topic in this field.

I have a question regarding the implementation of the short-history ablation (e.g., "10-step" history).
In your Appendix, you mention that "the hidden state was reset every 10 steps."
Does this mean that the hidden state is zero-initialized every 10 steps (i.e., set to all zeros),
or is it detached from the computation graph while keeping its value?
In other words, is the model's memory truly cleared every 10 steps, or is the gradient flow interrupted?

Additionally, upon checking the implementation,
I found the following code, which seems to detach the hidden state at every chunk:

/// code ///
    def training_step(self, batch, batch_idx):
…
        pred_action, self.hiddens = self.policy.step(concat_lowdim, rgb, self.hiddens)
        self.hiddens = [
            ((c.detach() if c is not None else None), (s.detach() if s is not None else None))
            for (c, s) in self.hiddens
        ]
…
        loss = F.mse_loss(pred_action, actions)
        self.manual_backward(loss)
…
        return loss
///////////

From this, I assume that both the full-history and 10-step settings share the same chunk-wise detach operation,
and that in the 10-step case, there is an additional zero-initialization every 10 steps.
Is this understanding correct?

If so, could you please let me know the motivation for detaching the hidden state at every chunk?
Do you think that as long as the hidden state is carried over,
the model can reliably retain long-term dependencies even if gradients are detached?

I would appreciate clarification on this point, as it is important for understanding this model.

**Comment by yulinzhouZYL** [comment]
Hi @y-ito-1985,

Thank you so much for your kind words and for your incredibly detailed and insightful question. This is a fantastic point and gives us a great opportunity to clarify our implementation details, which are crucial for understanding the model's behavior.

You are correct in your analysis, but there seems to be a slight misunderstanding regarding which code corresponds to which experiment. Let me clarify the two different cases (the main full-history model vs. the 10-step ablation).

1. Full-History Model (The code currently in the repository)
The code snippet you've found is from our main full-history model. Your understanding of the .detach() operation is spot on—it severs the tensor from the current computation graph. However, its purpose here is not to reset or clear the hidden state's memory.

Instead, it's a standard and necessary technique for training recurrent models on very long sequences. This is effectively an implementation of Truncated Backpropagation Through Time (TBPTT) with a truncation length of 1.

Motivation for .detach(): Without detaching the hidden state, each manual_backward(loss) call would attempt to backpropagate gradients not only through the current step's network but all the way back through every preceding step in the trajectory history. For long trajectories (400+ steps), this is computationally infeasible and would lead to a massive memory overflow.

Information Flow vs. Gradient Flow: By detaching the hidden state, we ensure that the gradient flow is truncated at each training step. However, the information flow is preserved. The actual value of the hidden state, which contains the compressed history, is passed forward to the next step.

So, the key distinction is: The forward pass is stateful across the entire trajectory (information is carried), while the backward pass is truncated at each step (gradients are not).

2. Short-History ("10-step") Ablation
You are correct that the code for this specific ablation is not in the current repository, as it was used for comparison in the paper.

For that experiment, we implemented a hard reset, just as you hypothesized. The hidden state was explicitly re-initialized to zeros every 10 steps. So, to answer your question directly: yes, for the 10-step ablation, the model's memory was truly cleared periodically.

Summary and Your Final Question
This leads to your final, excellent question: "Do you think that as long as the hidden state is carried over, the model can reliably retain long-term dependencies even if gradients are detached?"

The answer is yes, and this is fundamental to why this approach works and is a standard paradigm for training modern recurrent networks. The model learns to encode and maintain useful long-term information in its hidden state because that information is essential for minimizing the loss at future steps. Even though the gradient from step t doesn't flow back to step t-100, the loss at step t is calculated using h_t-1, which does contain information propagated from step t-100. Therefore, the optimization process forces the network to learn a state transition function that preserves history effectively.

To summarize:

The posted code is for the full-history model.

The .detach() operation in that code is for managing gradient flow (a form of TBPTT), not for resetting memory. The history is fully preserved and carried forward in the hidden state's value.

The 10-step ablation model (not in the repo) did use a hard reset to zeros every 10 steps to intentionally clear its memory for the ablation study.

I hope this clarifies the implementation and the motivation behind the design. Your deep dive into the code and thoughtful questions are greatly appreciated! Please feel free to ask if anything is unclear or if you have any more questions.

Best regards.

**Comment by y-ito-1985** [comment]
From here on, specifically to the Full-History model.

Thanks to your explanation, I now understand the following:
- The belief (hidden state) is always propagated forward (unless the episode changes).
- Gradients are truncated at each chunk.
- If gradients are not truncated, computational cost increases significantly.

However, I'm still not fully convinced by the idea that the model actively learns
to preserve context in order to minimize future loss.

Rather than saying the model learns to preserve context through chunk-wise detachment,
it seems more accurate to say that context naturally persists due to the structure of the one-step window shift.

Because gradients are detached,
the model cannot adjust the influence (weights) of past data to minimize future loss.

That said, the fact that Mamba-based sequence models can still capture long-term dependencies
even when gradients are detached at each chunk is a valuable insight.


I hope you don't mind—just one more thing I'd like to ask.

Instead of choosing between full-history (too long) or single-chunk (too short?) detachment,
I wonder if it might be beneficial to avoid detaching for a few chunks (with randomly selected durations).
This should be a fairly obvious idea.

I also wondered whether the reason for not doing so is based on the insight
that overly considering context might exacerbate the distribution shift problem commonly seen in behavior cloning models,
and that the current system design was chosen to avoid that.
Or perhaps I'm overthinking it?

I apologize for the lengthy message, and thank you very much for your valuable comments.
It's not often that I get the chance to speak with someone who attends top-level conferences.

Best regards.

**Comment by yulinzhouZYL** [comment]
Hi @y-ito-1985,

Thank you for this fantastic and deeply insightful follow-up question. This gets to the very heart of how recurrent models learn from sequential data, and it's a pleasure to discuss it at this level.

On the Learning Mechanism
You've articulated a classic and very subtle point about training recurrent networks. You are absolutely correct that the gradient from the loss at step t does not directly update the weights based on the input from a distant past step like t-100.

However, the "active learning" of context preservation happens implicitly through the shared parameters of the Mamba model, which are updated at every single step.

Let's break it down: The loss L_t at the current step teaches the model's shared weights a lesson: "Given the summary of the past you had (h_{t-1}), how should you have transformed it and processed the current input (x_t) to produce the correct action now?" This pressure, applied sequentially across the entire trajectory, forces the model to learn a state transition function that is effective at preserving historically relevant information. If the model fails to pass along crucial information from step t-100 into h_{t-1}, it will likely perform poorly not just at step t, but also at t+1, t+2, and so on.

In essence, the model is constantly rewarded for being a good "history compressor" because a well-compressed history is essential for minimizing local, step-wise losses. So, while the credit assignment isn't direct via long-term gradients, it's enforced by the necessity of creating a useful state representation at every single moment.

On Backpropagating Through a Few Chunks
Your proposal to backpropagate through a few chunks before detaching (i.e., Truncated Backpropagation Through Time with a window size N > 1) is an excellent and very natural idea.

The primary reason we used a window of N=1 (detaching every chunk) is purely practical: computational cost and memory. Backpropagating through N steps requires storing the intermediate activations for all those steps, leading to a memory requirement that scales linearly with N. Given that our model processes high-dimensional visual inputs, even a small N (e.g., 10 or 20) would dramatically increase the memory footprint and could become prohibitively expensive on commodity hardware. (We provide a more detailed analysis of this computational cost in the final camera-ready version of our RA-L paper, which will be more comprehensive than the current arXiv preprint and is expected to be published in a couple of weeks for those interested in a deeper look.)

Theoretically, a larger N could allow the model to learn more complex temporal credit assignments. However, the choice of N=1 is a common and effective trade-off, especially for powerful architectures like Mamba that are already very good at propagating information through their hidden state.

On Distribution Shift
Your speculation about this choice potentially mitigating the distribution shift problem in BC is incredibly thoughtful. I must admit, it wasn't our primary motivation for the N=1 design, but it's a fascinating and very plausible hypothesis.

One could certainly argue that a shorter BPTT window acts as a form of regularization. It might force the model to learn a more robust policy that relies on a compact state representation, rather than overfitting to specific long-term action sequences found in the training data. This could, in turn, make the policy more resilient when it encounters slightly out-of-distribution states during rollout. This is a really interesting research question in itself!

Thank you again for this stimulating discussion. These are exactly the kinds of questions that help push research forward. Please don't apologize for the lengthy message—it was a pleasure to read. And likewise, it's a privilege to discuss these ideas with engaged members of the community like you.

Best regards.

**Comment by y-ito-1985** [comment]
Thanks!  
That was really helpful.  
Looking forward to your next work.

**Comment by yulinzhouZYL** [comment]
Hi @y-ito-1985,

I hope you've been doing well! I'm reaching out because I vividly remember our deep and inspiring discussion back in September regarding the .detach() operation and the Truncated Backpropagation Through Time (TBPTT) window size.

Back then, you insightfully pointed out the limitations of truncating the gradient at every single step (N=1) and suggested the benefits of backpropagating through longer chunks. I mentioned that computational and memory bottlenecks were the primary reasons holding us back from doing so.

I am thrilled to let you know that we have successfully overcome that bottleneck. I just pushed a major update to the repository featuring a fully parallelized training framework.

More importantly, this new version implements a true Full-Sequence BPTT scheme. We have completely eliminated the chunk-wise .detach() operation. The model now processes and backpropagates gradients across the entire trajectory sequence in parallel, capturing the "true full-history" dependency without the previous VRAM overflow issues, thanks to some new structural optimizations.

Your probing questions about the gradient flow were a huge motivation for me to rethink and ultimately solve this engineering challenge. We have integrated this upgraded architecture as an important engineering optimization in our next major extension, which we are preparing for submission to IEEE T-RO.We expect to release the related paper and further updates in the next 2-3 months.

Thank you again for being such a brilliant sounding board. I invite you to check out the new _par files in the repo if you're interested, and please keep an eye out for our upcoming T-RO work!

Best regards,
Yulin Zhou

**Comment by y-ito-1985** [comment]
Thank you for your message.

I understand that this is a implementation update that leverages Mamba's inherent parallel processing capabilities.
It is impressive that temporal dependencies across the entire episode length can be explicitly taken into account.
As with the original MTIL, it seems that you have established a deterministic BC learning method that accounts for long-term dependencies and is likely to inspire many researchers.

Since last autumn, inspired by the MTIL training framework, I have been exploring a hierarchical imitative policy that can handle non-markovian long-horizon tasks.
I also considered explicitly modeling interactions with beliefs about the state in order to suppress state-distribution shift.

Even if it may not be at the level of top journals like yours, I am also planning to publish a paper in the near future.

Best regard.

**Comment by yulinzhouZYL** [comment]
Hi @y-ito-1985,

Thank you so much for the incredibly kind words. I am truly humbled and thrilled to hear that the MTIL framework has served as an inspiration for your own research!

Your upcoming work sounds absolutely fascinating. Exploring a hierarchical imitative policy for non-Markovian long-horizon tasks is exactly the kind of next-level challenge our field needs to tackle. I am particularly intrigued by your approach to explicitly modeling interactions with state beliefs to suppress the distribution shift. It beautifully echoes our deep dive last September, and it directly targets one of the most notorious bottlenecks in Behavior Cloning. That is a brilliant and very promising direction.

Please don't downplay your work regarding publication venues! Meaningful insights, solid engineering, and innovative ideas are what truly advance our community, regardless of the journal or conference name.

I am genuinely excited to read your paper. Please do share the arXiv link or the paper with me once it's available online. I would love to learn from your approach!

Best of luck with the final stages of your manuscript preparation.

Best regards.
